### ISSUE TITLE
Data not read from source directory during init

### ISSUE THREAD

**Issue by zarelit** [twpayne/chezmoi#1336]
## Describe the bug

When `chezmoi init -S dir` is run and it contains a `.chezmoi.<format>.tmpl` data in `.chezmoidata.<format>` is not used

## To reproduce

1. Create a source directory e.g. `/tmp/test_bug`
2. Put the files described below
3. Run `chezmoi init -S /tmp/test_bug`
4. Look at ~/.config/chezmoi/chezmoi.yaml` and notice it's rendered without data

Content of the dir:
`.chezmoi.yaml.tmpl`
```
---
# Let chezmoi use the dotfiles repository
sourceDir: {{ .my_source_path | quote }}
```

`.chezmoidata.yaml`
```
---
my_source_path: /tmp/test_bug
```

## Expected behavior

The configuration file should contain the correct path.

All the other commands including `chezmoi execute-template --init -S /tmp/test_bug < .chezmoi.yaml.tmpl` and `chezmoi data -S /tmp/test_bug` will use the data in `.chezmoidata.yaml`

## Output of command with the `--verbose` flag

Output empty, status code 0

## Output of `chezmoi doctor`

After `chezmoi init -S /tmp/test_bug`
```console
$ chezmoi doctor
RESULT    CHECK                MESSAGE
warning   version              v2.0.12, built by nixpkgs
ok        os-arch              linux/amd64
ok        config-file          /home/zarel/.config/chezmoi/chezmoi.yaml is a file
error     source-dir           /home/zarel/.local/share/chezmoi: open /home/zarel/.local/share/chezmoi: no such file or directory
error     suspicious-entries   lstat /home/zarel/.local/share/chezmoi: no such file or directory
ok        dest-dir             /home/zarel is a directory
ok        shell                found /run/current-system/sw/bin/bash
ok        editor               found /run/current-system/sw/bin/vim
ok        git-cli              found /home/zarel/.nix-profile/bin/git, version 2.31.1
ok        merge-cli            found /run/current-system/sw/bin/vimdiff
ok        age-cli              age not found in $PATH
ok        gnupg-cli            found /run/current-system/sw/bin/gpg, version 2.2.27
info      1password-cli        op not found in $PATH
info      bitwarden-cli        bw not found in $PATH
info      gopass-cli           gopass not found in $PATH
ok        keepassxc-cli        found /run/current-system/sw/bin/keepassxc-cli, version 2.6.4
info      keepassxc-db         not set
info      lastpass-cli         lpass not found in $PATH
info      pass-cli             pass not found in $PATH
info      vault-cli            vault not found in $PATH
info      secret-cli           not set
```
n.b. it also happens on the latest release

## Additional context

This is needed to bootstrap chezmoi from a local checkout and keep tracking from there, i.e. I would like to have my dotfiles in `~/code/dotfiles` without the need to edit the source directory in the configuration at a later time

**Comment by twpayne** [comment]
Thank you for the very complete bug report - much appreciated!

> This is needed to bootstrap chezmoi from a local checkout and keep tracking from there, i.e. I would like to have my dotfiles in `~/code/dotfiles` without the need to edit the source directory in the configuration at a later time

You can do this by putting the following in your `.chezmoi.yaml.tmpl`:

```
sourceDir: {{ .chezmoi.sourceDir | quote }}
```

This will effectively "save" the initial `-S` value passed to `chezmoi init` in your config file.

> All the other commands including `chezmoi execute-template --init -S /tmp/test_bug < .chezmoi.yaml.tmpl` and `chezmoi data -S /tmp/test_bug` will use the data in `.chezmoidata.yaml`

`chezmoi execute-template --init` not behaving the same as `chezmoi init` in the handling of config file creation is definitely a bug, thank you! However, I'd like to handle this by making `chezmoi execute-template --init` behave like `chezmoi init`, not the other way round.

**Comment by twpayne** [comment]
> However, I'd like to handle this by making `chezmoi execute-template --init` behave like `chezmoi init`, not the other way round.

#1338 implements this, i.e. it makes `chezmoi execute-template --init` no longer read `.chezmoidata.<format>` files.

@zarelit does this `.chezmoi.yaml.tmpl` tip above resolve your use case?

**Comment by zarelit** [comment]
@twpayne Thanks for the trick, it solves the case of setting the sourcedir without passing through data.

What started this thing is the idea of having a sort of metaconfiguration i.e. I would like in the end to share the repository without sharing things such as the path of the keepassxc file or some private-but-not-secret data such as emails or list of logins. I feel like I'm slightly going out of the project scope but I cannot grasp where.

Anyway, my repo at the moment has this `.chezmoi.yaml.tmpl`:
```
---
# Let chezmoi use the dotfiles repository
sourceDir: {{ .dotfilesRepoPath | quote }}

# Use this keepassxc file to read every secret
keepassxc:
  database: {{ .passwordsDbPath | quote }}

encryption: age
age:
  passphrase: true
```

and a data file like this: `.chezmoidata.yaml`
```
---
passwordsDbPath: /home/zarel/path/to/passwords.kdbx
dotfilesRepoPath: /home/zarel/code/dotfiles

identities:
  git:
    name: David
    email: redacted
  git_bug:
    - id: redacted
      kind: token
      target: gitlab
      createtime: 1626461801
      salt: redacte
      value: redacted
      login: redacted
      base_url: https://gitlab.com

git:
  replacements:
    - from: https://bitbucket.org/
      to: "git@bitbucket.org:"

```

So in my case to avoid the hardcoded path to the keepassx file I can either: provide a chezmoi.yaml.tmpl example instead of the data file example and ignore it in the repo or try to read it from data (hence the issue).

I would like eventually to reach a point where I carry chezmoi and my keepassx db and get everything populated including SSH keys and GPG configurations but there are several impediments beyond this one:
- keepassx-cli does not yet support attachments so we can't take out a file from it
- age encryption needs an identity file or a passphrase so because of the first point the identity file cannot be provisioned by chezmoi itself and a passphrase must be stored out-of-band (in my head or in keepassx again)

**Comment by twpayne** [comment]
Some thoughts:

In my personal case, I have most of the logic of handling machine-to-machine differences in [my `.chezmoi.toml.tmpl`](https://github.com/twpayne/dotfiles/blob/master/.chezmoi.toml.tmpl). The end result is my chezmoi config file is only about 8-10 lines long, even though the template that its generated from is much longer.

In your case, you could consider generating `passwordsDbPath` in your `.chezmoi.yaml.tmpl`. For example, it might have a default value but then be changed based on the hostname, something like:

```
{{ $passwordsDbPath := joinPath .chezmoi.homeDir "path" "to" "passwords.db" -}}
{{ if eq .chezmoi.hostname "somehost" -}}
{{   $passwordsDbPath = "/some/other/path/to/passwords.db" -}}
{{ end -}}

---
data:
  passwordsDbPath: {{ $passwordsDbPath | quote }}
```

The effect of the above is to move the contents of the `.chezmoidata.<format>` files into `.chezmoi.yaml.tmpl`.

With respect to having a metaconfiguration file, I think it's much less complex and more robust to generate the dotfiles contents from a few high-level variables (e.g. the handful of boolean flags that I use) than to have some metaconfig and a complex template to generate a dotfile from structured data (e.g. the `.git.replacements` array in your example). See [this comment](https://github.com/twpayne/chezmoi/issues/1333#issuecomment-894812443) in particular and the discussion in #1333 in general.

**Comment by twpayne** [comment]
Note that this issue got closed because I merged #1338. Happy to continue discussing here.

**Comment by zarelit** [comment]
I'm in a very similar position to #1333, now that the fix is merged it is clearer that the data file will not be read during init and that the options should be directly rendered in the chezmoi configuration template file.
I consider it solved, thank you!
